### PR TITLE
Save teacher logits

### PR DESCRIPTION
**Pull request by jiqing-feng**
Run teacher model forward in advance and save teacher logits in training dataset, instead of running teacher model forward on every batch to get the teacher logits.

@echarlaix @sywangyi @XinyuYe-Intel @PenghuiCheng

### CODE REVIEW

File: optimum/intel/neural_compressor/trainer.py [hunk 1]
```diff
@@ -549,16 +582,13 @@ def compute_loss(self, model, inputs, return_outputs=False):
...
-            and self.agent.criterion is not None
-        ):
+        if (self.is_in_train and hasattr(self, "agent") and hasattr(self.agent, "criterion") \
+            and self.agent.criterion is not None):
             logits = self._get_logits(outputs)
             teacher_logits = inputs.pop("teacher_logits", None)
-            # Compute teacher model outputs
-            if teacher_logits is None:
+            if teacher_logits is not None:
+                teacher_logits = tuple(teacher_logits.transpose(1, 0))
```
Comment (XinyuYe-Intel) [inline optimum/intel/neural_compressor/trainer.py:590]: this way of dealing teacher_logits only suit for SQuAD task, will be wrong for GLUE tasks since they usually have only one logit instead of two logits like SQuAD.

Comment (jiqing-feng) [reply optimum/intel/neural_compressor/trainer.py]: I have fixed it by conditional statements. The operation of transpose will be implemented only if the task has two logits.

**Comment by jiqing-feng** [comment]
@echarlaix Hi, Ella. Could you please help to check if I can merge it into HuggingFace/main? Thanks

**Comment by jiqing-feng** [comment]
@echarlaix Hi, Ella. I have fixed the code style and all checks have passed. Could you help me merge it? Thanks.

File: examples/neural_compressor/question-answering/run_qa.py [hunk 2]
```diff
@@ -513,6 +543,84 @@ def prepare_train_features(examples):
...
+        assert data_args.pad_to_max_length, "to run teacher logits must open pad_to_max_length due to padding issue"
+        teacher_config = AutoConfig.from_pretrained(optim_args.teacher_model_name_or_path)
+        teacher_model = AutoModelForQuestionAnswering.from_pretrained(
+            optim_args.teacher_model_name_or_path,
+            from_tf=bool(".ckpt" in optim_args.teacher_model_name_or_path),
+            config=teacher_config,
+        )
+        teacher_model = QAModel_output_reshaped(teacher_model)
+        teacher_model = accelerator.prepare(teacher_model)
+        para_counter = lambda model: sum(p.numel() for p in model.parameters())
```
Comment (echarlaix) [inline examples/neural_compressor/question-answering/run_qa.py:571]: What do you think about 
```suggestion
        num_param = lambda model: sum(p.numel() for p in model.parameters())
```

**Comment by jiqing-feng** [comment]
@echarlaix  Hi, Ella. Thanks for your comments. I have fixed most of them except generate teacher logits under optim_args.apply_distillation because it changes the dataset. However, I tried another way to avoid reload teacher model and config. Would you please help me to review these changes and merge it if there are no problems. Thanks!

**Review by hshen14 — APPROVED** [review]
LGTM

File: examples/neural_compressor/question-answering/run_qa.py [hunk 3]
```diff
@@ -513,6 +545,87 @@ def prepare_train_features(examples):
...
+
+        def get_logits(teacher_model_qa, train_dataset):
+            logger.info("***** Getting logits of teacher model *****")
+            logger.info(f"  Num examples = {len(train_dataset) }")
+            logger.info(f"  Batch Size = {training_args.per_device_eval_batch_size }")
+            npy_file = os.path.join(
+                os.path.dirname(os.path.abspath(__file__)),
+                "{}.{}.npy".format(data_args.dataset_name, optim_args.teacher_model_name_or_path.replace("/", ".")),
+            )
+            if os.path.exists(npy_file):
```
Comment (echarlaix) [inline examples/neural_compressor/question-answering/run_qa.py:591]: I would be in favor of removing the saving / loading of the teacher logits as the training dataset could vary in length or be preprocessed differently from one run to another

**Comment by jiqing-feng** [comment]
> Thanks for the addition @jiqing-feng, I left a comment about the teacher logits saving/loading

@echarlaix  Hi, Ella. Thanks for your advice, I have fixed this issue. The teacher logits would not be saved / loaded as .npy file.

**Review by echarlaix — APPROVED** [review]
LGTM, thanks for the great addition @jiqing-feng
